metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/CachedFieldNameConverterFactory.java

```diff
@@ -0,0 +1,155 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.metron.elasticsearch.writer;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.apache.commons.lang.ClassUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.metron.common.configuration.writer.WriterConfiguration;
+import org.apache.metron.common.field.DeDotFieldNameConverter;
+import org.apache.metron.common.field.FieldNameConverter;
+import org.apache.metron.common.field.FieldNameConverters;
+import org.apache.metron.common.field.NoopFieldNameConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A {@link FieldNameConverterFactory} that is backed by a cache.
+ *
+ * <p>Each sensor type can use a different {@link FieldNameConverter} implementation.
+ *
+ * <p>The {@link WriterConfiguration} allows a user to define the {@link FieldNameConverter}
+ * that should be used for a given sensor type.
+ *
+ * <p>The {@link FieldNameConverter}s are maintained in a cache for a fixed period of time
+ * after they are created.  Once they expire, the {@link WriterConfiguration} is used to
+ * reload the {@link FieldNameConverter}.
+ *
+ * <p>The user can change the {@link FieldNameConverter} in use at runtime. A change
```

Comment:
I want to say that this is handled through the ConfigurationsUpdater classes already already. I believe the writer passes in an updated config on each call.

https://github.com/apache/metron/blob/master/metron-platform/metron-common/src/main/java/org/apache/metron/common/zookeeper/configurations/ConfigurationsUpdater.java

Reply:
Maybe I am misunderstanding you, but the ConfigurationsUpdater ensures that our configuration classes stay in-sync with Zk.  That's now what this class does. 

This maintains a `FieldNameConverter` instance for each source type that is then used by the `ElasticsearchWriter` to do the field name conversion.  Are you proposing a different way to do this?

Reply:
To help clarify, I could have implemented a different `FieldNameConverterFactory`; maybe one called `SimpleFieldNameConverterFactory`.  

This factory wouldn't even have to use a cache.  Every time `ElasticsearchWriter` calls `create(String sensorType, WriterConfiguration config)`, it could just instantiate a new instance of the right `FieldNameConverter`.  

But of course, I didn't do that because this occurs for every field in every message that is written to Elasticsearch.  The cache is there for performance.

I offer this only as an example to help clarify the purpose of this class.

Reply:
I think I'm just not fully grokking why we need additional caching here rather than `FieldNameConverters.DEDOT.get();`. It seems like the enum handles caching by virtue of the constructors in the enum itself? `NOOP(new NoopFieldNameConverter())`

Reply:
So you're saying, hey why not just...
1. Take out the caching and just have something like a `SimpleFieldNameConverterFactory`. This would translate the (sourceType, WriterConfiguration) to the right `FieldNameConverter`.
2. The caching would be handled by `FieldNameConverters` in the sense that they are effectively singletons.

I think that would work if all `FieldNameConverter` implementations are stateless.  And they all are stateless today, but I tried not to make that a constraint.  I actually tried to make this work whether they are stateless or not.  

But alas there is a bug in what I did because I need each call to FieldNameConverters.get() to return a new instance.

I'd be open to removing the cache, if its not needed and we think `FieldNameConverter`s would always remain stateless.

Reply:
I definitely appreciate the thinking around state though I think we can safely leave that out for now. This isn't client-facing, so refactoring later would have less consequence than something we expose via config or the UI.

I think you could probably use the enum as a factory directly and use FieldNameConverter converter = FieldNameConverters.valueOf(converterName).get() in ElasticsearchWriter line 80 directly. We do this a number of places as a combination strategy/factory pattern. 

https://github.com/apache/metron/blob/master/metron-platform/metron-common/src/main/java/org/apache/metron/common/utils/CompressionStrategies.java

https://github.com/apache/metron/blob/master/metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/bolt/UnifiedEnrichmentBolt.java#L194

I definitely get wanting to provide a factory, and I think you effectively get that with the enum you already created.

Reply:
I still need a place to house the (sourceType, WriterConfiguration) -> FieldNameConverter logic.  I will remove the cache mechanism though.

Reply:
I actually liked what you had with your enum pattern. I was imaging a very minor change to your code, like below. The other change would be having the default converter value provided by the config class, or making it an Optional<String> if you wanted the es writer to own setting the default:

```
public interface FieldNameConverter {
    String convert(String originalField);
}

public enum FieldNameConverters implements FieldNameConverter {

  NOOP(new NoopFieldNameConverter()),
  DEDOT(new DeDotFieldNameConverter());

  private FieldNameConverter converter;

  FieldNameConverters(FieldNameConverter converter) {
    this.converter = converter;
  }

  @Override
  public String convert(String originalField) {
    return converter.convert(originalField);
  }
}

public class ElasticsearchWriter implements BulkMessageWriter<JSONObject>, Serializable {
    write( ... ) {
        String converterType = config.getFieldNameConverter(sensorType);
        // or this -> String converterType = config.getFieldNameConverter(sensorType).orElse("DEDOT");
        // fetch the field name converter for this sensor type
        FieldNameConverter fieldNameConverter = FieldNameConverters.valueOf(converterType);
        final String indexPostfix = dateFormat.format(new Date());
        BulkRequestBuilder bulkRequest = client.prepareBulk();
        for(JSONObject message: messages) {
          JSONObject esDoc = new JSONObject();
          for(Object k : message.keySet()){
            copyField(k.toString(), message, esDoc, fieldNameConverter);
          }
          ...
        }
        ...
    }

    private void copyField(... FieldNameConverter fieldNameConverter) {
         String destinationFieldName = fieldNameConverter.convert(sourceFieldName);
        ...
    }
...
}

```

Reply:
The primary reason the logic was extracted into a separate class is because the `ElasticsearchWriter` cannot be easily unit tested as it stands today.  I ran into tons of library conflicts when trying to do that.    

Keeping it separate also will make it easier to push out similar logic to other writers in the future; HDFS, Solr, should that be needed.

Reply:
The prior commit 'b18c3bf' was just me fixing my original caching implementation, so at least I had a version of that with the bug fixed.

The latest commit 'cee3994' removes the cache and relies on the `FactoryNameConverters` class to hand out shared instances.  This is what we have discussed.

Reply:
Yes, agreed on the possible need for handling other writers. Is there a reason we wouldn't put that logic in the enum factory rather than creating another new class for it? It wouldn't even need to depend on the WriterConfiguration, actually.

```
public enum FieldNameConverters implements FieldNameConverter {

  NOOP(new NoopFieldNameConverter()),
  DEDOT(new DeDotFieldNameConverter());

  private FieldNameConverter converter;

  FieldNameConverters(FieldNameConverter converter) {
    this.converter = converter;
  }

  public FieldNameConverter create(String sensorType, Optional<String> converterName) {
    // default to the 'DEDOT' field name converter to maintain backwards compatibility
   return FieldNameConverters.valueOf(converterName.orElse("DEDOT"));
  }

  @Override
  public String convert(String originalField) {
    return converter.convert(originalField);
  }
}
```

I'm not sure I follow about the testing - do you mean integration testing? We have an ElasticsearchWriterTest class that @justinleet wrote - https://github.com/apache/metron/blob/master/metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/writer/ElasticsearchWriterTest.java

Reply:
> I'm not sure I follow about the testing

If you look closely at those tests, all that class tests is the response using `buildWriteReponse`.  It does not actually test the class using the `write` method.  When you try to do that, it blows up because of dependency issues.

Reply:
Ok, I see what you mean. Agreed, it would be good to have some tests around that `write` method at some point, but I think it's ok to keep that out of scope for this particular PR. We have a WriterBoltIntegrationTest, but it's only covering parsers right now. At a later time we could look at expanding it to cover multiple writers and topologies and provide hooks on those builders to enable one-time setup and teardown across a suite of tests.

Reply:
I can add the `create` to the `FieldNameConverters`.  That makes sense here.  Its going to be slightly different than what you have above, but is what you were going for.